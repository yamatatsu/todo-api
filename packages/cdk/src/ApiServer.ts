import {
  App,
  Stack,
  StackProps,
  aws_ec2,
  aws_lambda,
  aws_lambda_nodejs,
  aws_apigateway,
  aws_secretsmanager,
  Duration,
  aws_cognito,
} from "aws-cdk-lib";

type Props = StackProps & {
  codeEntry: string;
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  dbCredentialSecret: aws_secretsmanager.ISecret;
  proxyEndpoint: string;
  userPool: aws_cognito.IUserPool;
};

export class ApiServerStack extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    // TODO: lambda insight入れる
    const handler = new aws_lambda_nodejs.NodejsFunction(this, "Lambda", {
      entry: props.codeEntry,
      runtime: aws_lambda.Runtime.NODEJS_14_X,
      // TODO: aurora serverless やめて rds proxyつけたら10秒に戻す
      timeout: Duration.seconds(20),
      vpc: props.vpc,
      securityGroups: [props.securityGroup],
      bundling: {
        /**
         * esbuildの成果物にprismaのengineが含める必要がある
         * Respect for https://dev.to/prisma/bundling-prisma-with-the-cdk-using-aws-lambda-nodejs-2lkd
         */
        nodeModules: ["@prisma/client", "prisma"],
        commandHooks: {
          beforeBundling(inputDir: string, outputDir: string): string[] {
            return [];
          },
          beforeInstall(inputDir: string, outputDir: string) {
            return [`cp -R ../prisma ${outputDir}/`];
          },
          afterBundling(inputDir: string, outputDir: string): string[] {
            return [
              `cd ${outputDir}`,
              `yarn prisma generate`,
              `rm -rf node_modules/@prisma/engines`,
            ];
          },
        },
        /**
         * forceDockerBundling: これは default false であるが、下記理由により明示的に示す。
         * これが true のとき、もしくは esbuild が見つからないとき(下記URL参照)に、
         * NodejsFunction では docker を用いてビルドが行われる。
         * https://github.com/aws/aws-cdk/blob/4c8e938e01b87636390a4f04de63bcd4dfe44cf8/packages/@aws-cdk/aws-lambda-nodejs/lib/esbuild-installation.ts#L8-L31
         * dockerでビルドする場合、特に Mac M1 での実行結果と GitHub actions 上での実行結果とで、
         * 成果物のfingerprintがズレて snapshot test が落ちてしまう。
         * そのため、 forceDockerBundling を明示し、 package.json でも esbuild を明示的にインストールする。
         */
        forceDockerBundling: false,
      },
      environment: {
        NODE_ENV: "production",
        SECRET_NAME: props.dbCredentialSecret.secretName,
        DATABASE_ENDPOINT: props.proxyEndpoint,
      },
    });

    props.dbCredentialSecret.grantRead(handler);

    // const authorizer = new aws_apigateway.CognitoUserPoolsAuthorizer(
    //   this,
    //   "Authorizer",
    //   {
    //     cognitoUserPools: [],
    //     identitySource: "method.request.header.x-ta-token",
    //   }
    // );

    const restApi = new aws_apigateway.LambdaRestApi(this, "LambdaRestApi", {
      handler,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: aws_apigateway.MethodLoggingLevel.INFO,
        dataTraceEnabled: true,
        tracingEnabled: true,
      },
      // defaultMethodOptions: {
      //   authorizer,
      // },
    });
  }
}
