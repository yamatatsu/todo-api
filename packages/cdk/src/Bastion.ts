import { App, Stack, StackProps, aws_ec2, aws_iam } from "aws-cdk-lib";

type BastionProps = {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
} & StackProps;

export class BastionStack extends Stack {
  readonly bastion: aws_ec2.Instance;

  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const instanceType = new aws_ec2.InstanceType("t2.micro");
    const machineImage = new aws_ec2.AmazonLinuxImage({
      generation: aws_ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
    });

    // ec2 instance for bastion server.
    this.bastion = new aws_ec2.Instance(this, "Bastion", {
      instanceType: instanceType,
      machineImage: machineImage,
      vpc: props.vpc,
      vpcSubnets: { subnetGroupName: "application" },
    });

    this.bastion.addSecurityGroup(props.securityGroup);

    // for ssm session manager
    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: [
          "ssm:UpdateInstanceInformation",
          "ssmmessages:CreateControlChannel",
          "ssmmessages:CreateDataChannel",
          "ssmmessages:OpenControlChannel",
          "ssmmessages:OpenDataChannel",
        ],
        resources: ["*"],
      })
    );

    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: ["s3:GetEncryptionConfiguration"],
        resources: ["*"],
      })
    );
  }
}
