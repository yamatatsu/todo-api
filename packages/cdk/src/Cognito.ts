import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_cognito,
} from "aws-cdk-lib";

export class CognitoStack extends Stack {
  public readonly userPool: aws_cognito.IUserPool;

  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props);

    const userPool = new aws_cognito.UserPool(this, "UserPool", {
      selfSignUpEnabled: true,
      signInAliases: { username: true, email: true },
      customAttributes: {
        validated_email: new aws_cognito.StringAttribute({
          mutable: true,
        }),
      },
      passwordPolicy: {
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#-5112-memorized-secret-verifiers
        minLength: 8,
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#a3-complexity
        requireLowercase: false,
        requireUppercase: false,
        requireDigits: false,
        requireSymbols: false,
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new aws_cognito.UserPoolClient(this, "UserPoolClient", {
      userPool,
      authFlows: { userSrp: true },
    });

    this.userPool = userPool;
  }
}
