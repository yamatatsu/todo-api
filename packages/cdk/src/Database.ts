import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
};

export class DatabaseStack extends Stack {
  public readonly dbCredentialSecret: aws_secretsmanager.ISecret;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const dbCluster = new aws_rds.ServerlessCluster(this, "Aurora", {
      engine: aws_rds.DatabaseClusterEngine.AURORA,
      credentials: aws_rds.Credentials.fromGeneratedSecret("LambdaRdsProxy"),
      defaultDatabaseName: "tadb",
      vpc: props.vpc,
      vpcSubnets: { subnetType: aws_ec2.SubnetType.ISOLATED },
      securityGroups: [props.securityGroup],
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // production codeではないので、カジュアルに`!`使う
    this.dbCredentialSecret = dbCluster.secret!;
  }
}
