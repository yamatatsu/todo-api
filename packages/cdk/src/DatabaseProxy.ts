import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  cluster: aws_rds.IDatabaseCluster;
  secret: aws_secretsmanager.ISecret;
};

export class DatabaseProxyStack extends Stack {
  public readonly proxy: aws_rds.IDatabaseProxy;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const proxy = props.cluster.addProxy("RdsProxy", {
      vpc: props.vpc,
      secrets: [props.secret],
      securityGroups: [props.securityGroup],
    });
    // proxy.connections.allowFromAnyIpv4(aws_ec2.Port.tcp(3306));

    this.proxy = proxy;
  }
}
