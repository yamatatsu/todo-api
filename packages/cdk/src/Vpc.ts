import { App, Stack, StackProps, aws_ec2 } from "aws-cdk-lib";

export class VpcStack extends Stack {
  public readonly vpc: aws_ec2.IVpc;

  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props);

    const vpc = new aws_ec2.Vpc(this, "Vpc", {
      natGateways: 0,
      maxAzs: 2,
      // natGatewayProvider: aws_ec2.NatInstanceProvider.instance({
      //   instanceType: new aws_ec2.InstanceType("t2.nano"),
      // }),
      subnetConfiguration: [
        // {
        //   cidrMask: 24,
        //   name: "ingress",
        //   subnetType: aws_ec2.SubnetType.PUBLIC,
        // },
        {
          cidrMask: 24,
          name: "application",
          subnetType: aws_ec2.SubnetType.ISOLATED,
        },
        {
          cidrMask: 28,
          name: "rds",
          subnetType: aws_ec2.SubnetType.ISOLATED,
        },
      ],
    });

    new aws_ec2.InterfaceVpcEndpoint(this, "SecretManagerVpcEndpoint", {
      vpc: vpc,
      service: aws_ec2.InterfaceVpcEndpointAwsService.SECRETS_MANAGER,
    });

    this.vpc = vpc;
  }
}
