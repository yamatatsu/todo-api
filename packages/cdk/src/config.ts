const ENV_NAME = ["development", "production"] as const;
type EnvName = typeof ENV_NAME[number];

export function assertEnvName(str: string): asserts str is EnvName {
  if (!ENV_NAME.includes(str as EnvName)) {
    throw new Error(`Invalid EnvName has set: ${str}`);
  }
}

export const AccountId: Record<EnvName, string> = {
  development: "407172421073",
  production: "",
};
