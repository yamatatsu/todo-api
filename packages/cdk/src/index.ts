import path from "path";
import { App, Environment } from "aws-cdk-lib";
import { CognitoStack } from "./Cognito";
import { VpcStack } from "./Vpc";
import { DatabaseStack } from "./Database";
import { ApiServerStack } from "./ApiServer";
import { BastionStack } from "./Bastion";
import { AccountId, assertEnvName } from "./config";

const envName = process.env.ENV_NAME || "development";
assertEnvName(envName);

const app = new App();
const env: Environment = {
  region: "ap-northeast-1",
  account: AccountId[envName],
};

const cognito = new CognitoStack(app, `${envName}-TodoApi-Cognito`, { env });
const { vpc, dbSG, dbAccessSG } = new VpcStack(app, `${envName}-TodoApi-Vpc`, {
  env,
});
const database = new DatabaseStack(app, `${envName}-TodoApi-Database`, {
  vpc,
  securityGroup: dbSG,
  env,
});
new ApiServerStack(app, `${envName}-TodoApi-ApiServer`, {
  codeEntry: path.resolve(__dirname, "../../server/src/index.ts"),
  vpc,
  securityGroup: dbAccessSG,
  dbCredentialSecret: database.dbCredentialSecret,
  userPoolArn: cognito.userPoolArn,
  env,
});
new BastionStack(app, `${envName}-TodoApi-Bastion`, {
  vpc,
  securityGroup: dbAccessSG,
  env,
});
