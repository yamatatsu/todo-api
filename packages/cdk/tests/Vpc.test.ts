import { SynthUtils } from "@aws-cdk/assert";
import { App, Environment } from "aws-cdk-lib";
import { VpcStack } from "../src/Vpc";
import { AccountId } from "../src/config";

test("snapshot test", () => {
  const app = new App();
  const env: Environment = {
    region: "ap-northeast-1",
    account: AccountId["development"],
  };

  const target = new VpcStack(app, "Target", { env });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
