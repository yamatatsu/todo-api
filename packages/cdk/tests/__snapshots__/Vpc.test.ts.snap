// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "SecretManagerVpcEndpoint708CAAF9": Object {
      "Properties": Object {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "SecretManagerVpcEndpointSecurityGroupB2F67559",
              "GroupId",
            ],
          },
        ],
        "ServiceName": "com.amazonaws.ap-northeast-1.secretsmanager",
        "SubnetIds": Array [
          Object {
            "Ref": "VpcapplicationSubnet1Subnet8B68205B",
          },
          Object {
            "Ref": "VpcapplicationSubnet2Subnet29670962",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "SecretManagerVpcEndpointSecurityGroupB2F67559": Object {
      "Properties": Object {
        "GroupDescription": "Target/SecretManagerVpcEndpoint/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": Object {
              "Fn::GetAtt": Array [
                "Vpc8378EB38",
                "CidrBlock",
              ],
            },
            "Description": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "from ",
                  Object {
                    "Fn::GetAtt": Array [
                      "Vpc8378EB38",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Target/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcapplicationSubnet1RouteTableAssociation75D5245D": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcapplicationSubnet1RouteTableF03E8296",
        },
        "SubnetId": Object {
          "Ref": "VpcapplicationSubnet1Subnet8B68205B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcapplicationSubnet1RouteTableF03E8296": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/applicationSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcapplicationSubnet1Subnet8B68205B": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/applicationSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcapplicationSubnet2RouteTable17378AED": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/applicationSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcapplicationSubnet2RouteTableAssociationF25110CC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcapplicationSubnet2RouteTable17378AED",
        },
        "SubnetId": Object {
          "Ref": "VpcapplicationSubnet2Subnet29670962",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcapplicationSubnet2Subnet29670962": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/applicationSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcrdsSubnet1RouteTableAssociationE01668F2": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcrdsSubnet1RouteTableE62E4ED6",
        },
        "SubnetId": Object {
          "Ref": "VpcrdsSubnet1SubnetC2926CEA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcrdsSubnet1RouteTableE62E4ED6": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/rdsSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcrdsSubnet1SubnetC2926CEA": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.2.0/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "rds",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/rdsSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcrdsSubnet2RouteTable3E531D9B": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/rdsSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcrdsSubnet2RouteTableAssociation25A7BD68": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcrdsSubnet2RouteTable3E531D9B",
        },
        "SubnetId": Object {
          "Ref": "VpcrdsSubnet2Subnet70A835C8",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcrdsSubnet2Subnet70A835C8": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.2.16/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "rds",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "Target/Vpc/rdsSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
